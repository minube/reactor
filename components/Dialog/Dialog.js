import {
  array, bool, func, node, number, object, oneOfType, string,
} from 'prop-types';
import React, { PureComponent } from 'react';
import {
  KeyboardAvoidingView, SafeAreaView, ScrollView, View,
} from 'react-native';

import { ENV, LAYOUT, THEME } from '../../common';
import Button from '../Button';
import Motion from '../Motion';
import Text from '../Text';
import styles from './Dialog.style';

const { IS_NATIVE } = ENV;
const { COLOR, MOTION } = THEME;

export default class Dialog extends PureComponent {
  static propTypes = {
    background: bool,
    children: node,
    highlight: bool,
    onClose: func,
    reverse: bool,
    style: oneOfType([array, number, object]),
    styleContainer: oneOfType([array, number, object]),
    title: string,
    visible: bool,
  };

  static defaultProps = {
    background: true,
    children: undefined,
    highlight: undefined,
    onClose: undefined,
    reverse: false,
    style: [],
    styleContainer: [],
    title: undefined,
    visible: false,
  };

  state = {
    scroll: false,
  };

  _onScroll = () => {
    this.setState({ scroll: true });
  }

  render() {
    const {
      _onScroll,
      props: {
        background, children, highlight, reverse, onClose, style, styleContainer, title, visible,
      },
      state: { scroll },
    } = this;
    const { VIEWPORT: { H, PORTRAIT } } = LAYOUT;
    let translateY = 0;
    if (!visible) translateY = reverse ? -H : H;

    return (
      <Motion
        delay={visible ? 0 : MOTION.DURATION}
        pointerEvents={((background || IS_NATIVE) && visible) ? 'auto' : 'none'}
        style={styles.container}
        timeline={[{ property: 'opacity', value: visible ? 1 : 0 }]}
      >
        <SafeAreaView style={[styles.safeArea, background && styles.background, styleContainer]}>
          <KeyboardAvoidingView
            behavior={IS_NATIVE ? 'padding' : undefined}
            style={[
              {
                maxHeight: PORTRAIT ? '100%' : '90%',
                minWidth: PORTRAIT ? 320 : '33%',
                maxWidth: PORTRAIT ? '100%' : '66%',
                position: 'absolute',
              },
              { backgroundColor: COLOR.TRANSPARENT },
            ]}
          >
            <Motion
              delay={visible ? MOTION.DURATION : 0}
              duration={MOTION.DURATION}
              pointerEvents="auto"
              type="timing"
              timeline={[{ property: 'translateY', value: translateY }]}
            >
              <View style={[styles.frame, style]}>
                <View style={styles.header}>
                  { title && (
                    <Text
                      color={highlight ? COLOR.WHITE : undefined}
                      headline
                      level={6}
                      numberOfLines={1}
                      style={styles.title}
                    >
                      {title}
                    </Text>
                  )}
                  { onClose && (
                    <Button
                      contained={false}
                      color={highlight ? undefined : COLOR.TEXT}
                      icon={highlight ? 'closeContrast' : 'close'}
                      onPress={onClose}
                      rounded
                    />
                  )}
                </View>
                <ScrollView onScroll={title ? _onScroll : undefined} style={[styles.children, scroll && styles.scroll]}>
                  {children}
                </ScrollView>
              </View>
            </Motion>
          </KeyboardAvoidingView>
        </SafeAreaView>
      </Motion>
    );
  }
}
