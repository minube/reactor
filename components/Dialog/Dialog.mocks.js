import React from 'react';
import Text from '../Text';
import { THEME } from '../../common';

const CHILDREN = <Text>lorem ipsum...</Text>;
const STYLE = { backgroundColor: THEME.COLOR.PRIMARY };
const STYLE_CONTAINER = { backgroundColor: THEME.COLOR.SECONDARY, height: '50%' };
const TITLE = 'Lorem ipsum...';

export default {
  CHILDREN,
  STYLE,
  STYLE_CONTAINER,
  TITLE,
};
