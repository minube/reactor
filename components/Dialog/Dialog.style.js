import { Platform, StyleSheet } from 'react-native';

import { STYLE, THEME } from '../../common';

const {
  BORDER_RADIUS, COLOR, FONT, OFFSET,
} = THEME;

export default StyleSheet.create({
  container: StyleSheet.flatten([
    STYLE.CENTERED,
    {
      backgroundColor: 'rgba(0,0,0, 0.5)',
      height: '100%',
      left: 0,
      position: Platform.OS === 'web' ? 'fixed' : 'absolute',
      top: 0,
      width: '100%',
      zIndex: 2,
    },
  ]),

  frame: StyleSheet.flatten([
    STYLE.SHADOW,
    {
      backgroundColor: COLOR.WHITE,
      borderRadius: BORDER_RADIUS,
      minWidth: '35%',
      maxHeight: '65%',
    },
  ]),

  title: {
    fontSize: FONT.SIZE.LARGE,
    padding: OFFSET,
  },

  content: {
    flex: 1,
    paddingHorizontal: OFFSET,
  },

  scroll: {
    borderBottomColor: COLOR.DIVIDER,
    borderBottomWidth: 1,
    borderTopColor: COLOR.DIVIDER,
    borderTopWidth: 1,
  },

  footer: StyleSheet.flatten([
    STYLE.ROW,
    {
      alignItems: 'flex-end',
      alignContent: 'flex-end',
      justifyContent: 'flex-end',
      padding: OFFSET,
    },
  ]),
});
