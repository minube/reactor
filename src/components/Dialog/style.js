import Color from 'color';
import { Platform, StyleSheet } from 'react-native';

import { THEME } from '../../common';

const {
  COLOR, FONT, LAYOUT, OFFSET, STYLE,
} = THEME;

export default StyleSheet.create({
  container: {
    backgroundColor: 'rgba(0,0,0, 0.5)',
    height: '100%',
    left: 0,
    position: Platform.OS === 'web' ? 'fixed' : 'absolute',
    top: 0,
    width: '100%',
    zIndex: 2,
    ...STYLE.CENTERED,
  },

  frame: {
    backgroundColor: COLOR.WHITE,
    minWidth: LAYOUT.DIALOG_MIN_WIDTH,
    maxHeight: '65%',
    ...STYLE.SHADOW,
  },

  title: {
    fontSize: FONT.SIZE.LARGE,
    fontWeight: FONT.WEIGHT.BOLD,
    padding: OFFSET,
  },

  content: {
    flex: 1,
    paddingHorizontal: OFFSET,
  },

  scroll: {
    borderBottomColor: COLOR.DIVIDER,
    borderBottomWidth: 1,
    borderTopColor: COLOR.DIVIDER,
    borderTopWidth: 1,
  },

  footer: {
    alignItems: 'flex-end',
    alignContent: 'flex-end',
    justifyContent: 'flex-end',
    padding: OFFSET,
    ...STYLE.ROW,
  },
});
