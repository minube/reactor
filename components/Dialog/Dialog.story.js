import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean } from '@storybook/addon-knobs/react';

import ResponsiveView from '../ResponsiveView';
import Dialog from './Dialog';
import Text from '../Text';

storiesOf('🛠 Dialog', module)
  .addWithJSX('default', () => (
    <Dialog  />
  ))
  .addWithJSX('visible', () => (
    <Dialog visible />
  ))
  .addWithJSX('background (disabled)', () => (
    <Dialog background={false} visible />
  ))
  .addWithJSX('title', () => (
    <Dialog title="Hello world" visible />
  ))
  .addWithJSX('⚡ onClose', () => (
    <Dialog title="Hello world" onClose={action('Dialog.onClose()')} visible />
  ))
  .addWithJSX('⚡ onSubmit', () => (
    <Dialog title="Hello world" onSubmit={action('Dialog.onSubmit()')} visible />
  ))
  .addWithJSX('style', () => (
    <Dialog title="Hello world" style={{ backgroundColor: 'orange', width: 512, height: 320 }} visible />
  ))
  .addWithJSX('styleContainer', () => (
    <Dialog
      title="Hello world"
      styleContainer={{ padding: 32, alignItems: 'flex-start', justifyContent: 'flex-end' }}
      visible
    />
  ))
  .addWithJSX('🏀 Playground', () => (
    <ResponsiveView>
      <Dialog
        background={boolean('background', true)}
        title={text('title', 'Title')}
        onClose={action('Dialog.onClose()')}
        onSubmit={action('Dialog.onSubmit()')}
        visible={boolean('visible', true)}
      >
        <Text>Example of children</Text>
      </Dialog>
    </ResponsiveView>
  ));
