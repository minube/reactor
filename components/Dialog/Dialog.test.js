import React from 'react';
import renderer from 'react-test-renderer';

import Dialog from './Dialog';
import MOCKS from './Dialog.mocks';

const EVENT = () => {};

describe('<Dialog>', () => {
  it('renders', () => {
    const tree = renderer.create(<Dialog />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {visible}', () => {
    const tree = renderer.create(<Dialog visible />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {reverse}', () => {
    const tree = renderer.create(<Dialog visible reverse />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {title}', () => {
    const tree = renderer.create(<Dialog title={MOCKS.TITLE} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {children}', () => {
    const tree = renderer.create(<Dialog>{MOCKS.CHILDREN}</Dialog>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {background}', () => {
    const tree = renderer.create(<Dialog background={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {style}', () => {
    const tree = renderer.create(<Dialog style={MOCKS.STYLE} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {styleContainer}', () => {
    const tree = renderer.create(<Dialog styleContainer={MOCKS.STYLE_CONTAINER} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {onClose}', () => {
    const tree = renderer.create(<Dialog onClose={EVENT} />).toJSON();
    expect(tree).toMatchSnapshot();
    // @TODO: Test <Dialog /> {onClose}
  });

  it('when {highlight}', () => {
    const tree = renderer.create(<Dialog title={MOCKS.TITLE} onClose={EVENT} highlight />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
