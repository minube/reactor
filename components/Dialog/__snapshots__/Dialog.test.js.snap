// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog> renders 1`] = `
<View
  collapsable={undefined}
  pointerEvents="none"
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "backgroundColor": "rgba(0,0,0,0.5)",
      "height": "100%",
      "justifyContent": "center",
      "left": 0,
      "opacity": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <View
    collapsable={undefined}
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 3,
        "maxHeight": "100%",
        "minWidth": "66%",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.26,
        "shadowRadius": 5,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
      }
    }
  >
    <View>
      <RCTScrollView
        onScroll={[Function]}
        style={
          Array [
            Object {
              "paddingHorizontal": 16,
            },
            false,
          ]
        }
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;

exports[`<Dialog> when {title} 1`] = `
<View
  collapsable={undefined}
  pointerEvents="none"
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "backgroundColor": "rgba(0,0,0,0.5)",
      "height": "100%",
      "justifyContent": "center",
      "left": 0,
      "opacity": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <View
    collapsable={undefined}
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 3,
        "maxHeight": "100%",
        "minWidth": "66%",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.26,
        "shadowRadius": 5,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={undefined}
        style={
          Array [
            Object {
              "color": "#4a4a4a",
              "fontFamily": undefined,
            },
            false,
            false,
            Object {
              "fontWeight": "bold",
            },
            false,
            Object {
              "fontSize": 16,
              "lineHeight": 20.8,
            },
            false,
            false,
            false,
            false,
            false,
            false,
            false,
            false,
            Object {
              "fontSize": 20,
              "padding": 16,
            },
          ]
        }
      >
        Hello world
      </Text>
      <RCTScrollView
        onScroll={[Function]}
        style={
          Array [
            Object {
              "paddingHorizontal": 16,
            },
            false,
          ]
        }
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;

exports[`<Dialog> when {visible} 1`] = `
<View
  collapsable={undefined}
  pointerEvents="auto"
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "backgroundColor": "rgba(0,0,0,0.5)",
      "height": "100%",
      "justifyContent": "center",
      "left": 0,
      "opacity": 1,
      "position": "absolute",
      "top": 0,
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <View
    collapsable={undefined}
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 3,
        "maxHeight": "100%",
        "minWidth": "66%",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.26,
        "shadowRadius": 5,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <View>
      <RCTScrollView
        onScroll={[Function]}
        style={
          Array [
            Object {
              "paddingHorizontal": 16,
            },
            false,
          ]
        }
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;
